### PR TITLE
[ME] Native Types for Postgres

### PR DESCRIPTION
**Pull request by do4gr**

### CODE REVIEW

File: libs/datamodel/connectors/sql-datamodel-connector/src/postgres_datamodel_connector.rs [hunk 1]
```diff
@@ -16,9 +15,6 @@ const DECIMAL_TYPE_NAME: &str = "Decimal";
 const NUMERIC_TYPE_NAME: &str = "Numeric";
```
Comment (do4gr) [inline libs/datamodel/connectors/sql-datamodel-connector/src/postgres_datamodel_connector.rs:15]: Decimal is also an alias that does not roundtrip

File: libs/datamodel/connectors/datamodel-connector/src/lib.rs [hunk 2]
```diff
@@ -35,6 +35,12 @@ pub trait Connector: Send + Sync {
     /// This powers the use of native types for QE + ME.
     fn parse_native_type(&self, name: &str, args: Vec<String>) -> Result<NativeTypeInstance, ConnectorError>;
 
+    /// This function is used in ME for error messages
+    fn render_native_type(&self, native_type: serde_json::Value) -> String {
+        let instance = self.introspect_native_type(native_type).unwrap(); //todo
```
Comment (tomhoule) [inline libs/datamodel/connectors/datamodel-connector/src/lib.rs:40]: is this likely to crash?

File: libs/datamodel/core/tests/types/postgres_native_types.rs [hunk 3]
```diff
@@ -1,21 +1,21 @@
...
+//         format!(
+//             "Sequential native type {} of Postgres must not have a static default value.",
+//             type_name
+//         )
+//     }
+//
+//     for tpe in &["SmallSerial", "Serial", "BigSerial"] {
+//         test_native_types_with_field_attribute_support(tpe, "Int", "default(4)", &error_msg(tpe), POSTGRES_SOURCE);
+//     }
+// }
```
Comment (tomhoule) [inline libs/datamodel/core/tests/types/postgres_native_types.rs:18]: Do you think we can delete this, or should we leave it in until we've had the opportunity to discuss serial types with product?

File: migration-engine/connectors/sql-migration-connector/Cargo.toml [hunk 4]
```diff
@@ -7,6 +7,7 @@ version = "0.1.0"
 [dependencies]
 datamodel = {path = "../../../libs/datamodel/core"}
 datamodel-connector = {path = "../../../libs/datamodel/connectors/datamodel-connector"}
+sql-datamodel-connector = {path = "../../../libs/datamodel/connectors/sql-datamodel-connector"}
```
Comment (tomhoule) [inline migration-engine/connectors/sql-migration-connector/Cargo.toml:10]: I think it's unfortunate to have three imports for what is conceptually one thing (datamodel), but I don't have an alternative in mind yet, so let's go with this.

File: migration-engine/migration-engine-tests/tests/migration_tests.rs [hunk 5]
```diff
@@ -2752,6 +2752,35 @@ async fn a_table_recreation_with_noncastable_columns_should_trigger_warnings(api
...
+//             id      Int @id
+//             float   DateTime
+//         }
+//     "#;
+//
+//     api.schema_push(dm2).send().await?;
+//     // .assert_warnings(&["You are about to alter the column `float` on the `Blog` table. The data in that column will be cast from `String` to `Float`. This cast may fail and the migration will stop. Please make sure the data in the column can be cast.".into()])?;
+//
+//     Ok(())
+// }
```
Comment (tomhoule) [inline migration-engine/migration-engine-tests/tests/migration_tests.rs:2782]: I can pick that one up next PR

File: migration-engine/connectors/sql-migration-connector/src/sql_schema_differ/sql_schema_differ_flavour/mysql.rs [hunk 6]
```diff
@@ -51,6 +51,10 @@ impl SqlSchemaDifferFlavour for MysqlFlavour {
         None
     }
 
+    //split into enum changes
+    //family changes
+    //native type changes
```
Comment (tomhoule) [inline migration-engine/connectors/sql-migration-connector/src/sql_schema_differ/sql_schema_differ_flavour/mysql.rs:56]: are these comments still relevant?

File: migration-engine/connectors/sql-migration-connector/src/sql_destructive_change_checker.rs [hunk 7]
```diff
@@ -179,6 +179,8 @@ impl SqlMigrationConnector {
                                 );
                             }
 
+                            // can this just use the flavour??
+                            // todo pull up type rendering and native type check here
```
Comment (tomhoule) [inline migration-engine/connectors/sql-migration-connector/src/sql_destructive_change_checker.rs:183]: :+1:

File: migration-engine/connectors/sql-migration-connector/src/sql_destructive_change_checker.rs [hunk 8]
```diff
@@ -207,6 +209,7 @@ impl SqlMigrationConnector {
                                         );
                                     }
                                 }
+                                //todo why no precise type warnings here if native types are on?
```
Comment (tomhoule) [inline migration-engine/connectors/sql-migration-connector/src/sql_destructive_change_checker.rs:212]: :+1:

Comment (tomhoule) [inline migration-engine/connectors/sql-migration-connector/src/sql_destructive_change_checker.rs:212]: :+1:

File: migration-engine/connectors/sql-migration-connector/src/flavour/postgres.rs [hunk 9]
```diff
@@ -11,11 +12,18 @@ use user_facing_errors::{
 };
 
 #[derive(Debug)]
-pub(crate) struct PostgresFlavour(pub(crate) PostgresUrl);
+pub(crate) struct PostgresFlavour {
+    pub(crate) url: PostgresUrl,
+    features: BitFlags<MigrationFeature>,
```
Comment (tomhoule) [inline migration-engine/connectors/sql-migration-connector/src/flavour/postgres.rs:17]: We talked about it yesterday, but I'm not sure why this shows up in the diff. Let's see if it's still there after a rebase.

File: migration-engine/connectors/sql-migration-connector/src/sql_destructive_change_checker/destructive_change_checker_flavour/postgres.rs [hunk 10]
```diff
@@ -101,6 +117,7 @@ impl DestructiveChangeCheckerFlavour for PostgresFlavour {
                 step_index,
             )
         } else {
+            //todo this is probably due to a not castable type change. we should give that info in the warning
```
Comment (tomhoule) [inline migration-engine/connectors/sql-migration-connector/src/sql_destructive_change_checker/destructive_change_checker_flavour/postgres.rs:120]: :+1:

I think the right way to approach these kinds of issues is to have a concrete case where we could render a better error message, put that in an issue and treat it as an improvement. It's hard to think about error ergonomics when you are in the middle of the code with all the case analysis.

File: migration-engine/migration-engine-tests/tests/native_types/postgres.rs [hunk 11]
```diff
@@ -0,0 +1,1093 @@
...
+        api.schema_push(&dm2).send().await?.assert_green()?;
+
+        //second assertions
+        api.assert_schema().await?.assert_table("A", |table| {
+            next_assertions.iter().fold(
+                table.assert_column_count(next_assertions.len() + 1),
+                |acc, (name, expected)| {
+                    acc.and_then(|table| table.assert_column(name, |c| c.assert_native_type(expected, &connector)))
+                },
+            )
```
Comment (tomhoule) [inline migration-engine/migration-engine-tests/tests/native_types/postgres.rs:869]: nice!

**Review by tomhoule — APPROVED** [review]
lgtm!

File: libs/datamodel/core/tests/types/postgres_native_types.rs (hunk 3 above)
Comment (do4gr) [reply libs/datamodel/core/tests/types/postgres_native_types.rs:18]: Yeah, wanted to leave it in due to that reason

File: libs/datamodel/connectors/datamodel-connector/src/lib.rs (hunk 2 above)
Comment (do4gr) [reply libs/datamodel/connectors/datamodel-connector/src/lib.rs:40]: Probably not, I think the native types json here should already have passed validation no matter where it is coming from
I'll remove the todo

File: migration-engine/connectors/sql-migration-connector/src/sql_schema_differ/sql_schema_differ_flavour/mysql.rs (hunk 6 above)
Comment (do4gr) [reply migration-engine/connectors/sql-migration-connector/src/sql_schema_differ/sql_schema_differ_flavour/mysql.rs:56]: these are just notes for the mysql work. My idea was to separate the differ logic into subcases to make it more palatable. For SQLServer that's just diffing columntypechanges vs native type changes. For Postgres lists are an additional concern. For MySQL its enums. You can remove these.

File: migration-engine/migration-engine-tests/tests/native_types/postgres.rs (hunk 11 above)
Comment (do4gr) [reply migration-engine/migration-engine-tests/tests/native_types/postgres.rs:869]: Hehe, yeah I thought you'd like that one

**Pull request merged**